### ISSUE TITLE
Reverse YouTube Playlist Sort Order

### ISSUE THREAD

**Issue by crxm** [ZeroQI/Absolute-Series-Scanner#326]
**Is your feature request related to a problem? Please describe.**
Episode ordering for some YouTube playlists can be an issue as new episodes may be added to existing playlists in reverse order.

Example: https://www.youtube.com/playlist?list=PL290QH8A1p-RHNIH80ZdTaKCYx8mKD01o

**Describe the solution you'd like**
It would be nice to have an option to determine the episode rank based on the reverse order of certain playlists. Ideally, this would be triggered by an optional flag in the folder name, if possible.

I am currently testing this change as a workaround:
`json_full_items = Dict(json_full, 'items')
          for rank, video in enumerate(json_full_items or {}, start=1):
            VideoID = video['snippet']['resourceId']['videoId']
            if VideoID and VideoID in file.decode('utf-8'):
              reversed_rank = len(json_full_items) - (rank - 1)
              Log.info('[{}] rank: {:>3} in file: {}'.format(VideoID, reversed_rank, file))
              add_episode_into_plex(media, os.path.join(root, path, file), root, path, folder_show, int(folder_season if folder_season is not None else 1), reversed_rank, video['snippet']['title'].encode('utf8'), "", reversed_rank, 'YouTube', tvdb_mapping, unknown_series_length, offset_season, offset_episode, mappingList)
              break`

(Sorry about the formatting. I'm not sure how to fix it here.)

**Describe alternatives you've considered**
youtube-dl is able to provide the desired playlist index for situations like this by using the --playlist-reverse flag, but that does not appear to affect episode numbering in this case.

**Additional context**
N/A

Thank you for your work on this project and for sharing it!

**Comment by ZeroQI** [comment]
<> on the code
```
json_full_items = Dict(json_full, 'items')
for rank, video in enumerate(json_full_items or {}, start=1):
  VideoID = video['snippet']['resourceId']['videoId']
  if VideoID and VideoID in file.decode('utf-8'):
    reversed_rank = len(json_full_items) -rank + 1
    Log.info('[{}] rank: {:>3} in file: {}'.format(VideoID, reversed_rank, file))
    add_episode_into_plex(media, os.path.join(root, path, file), root, path, folder_show, int(folder_season if folder_season is not None else 1), reversed_rank, video['snippet']['title'].encode('utf8'), "", reversed_rank, 'YouTube', tvdb_mapping, unknown_series_length, offset_season, offset_episode, mappingList)
    break

```
it's annoying when the playlist order change with new releases, which is why i used the file date on disk to mitigate that
tried undocumented youtube2 forced id? it will create a season per year though

**Comment by crxm** [comment]
Using the file date on disk seems like it might work in most cases, but I ran into an odd issue with that at one point as I tried out a few options. I'm not sure which date youtube-dl uses when it downloads videos, but for a few videos in the playlist I linked earlier the file date on disk was later than the original release date. This caused the episode numbers to be incorrect. I assume that the uploaded had reuploaded the video at some point at youtube-dl was now using that date as the file date instead of the original release date. Just a guess though.

I hadn't seen anything about a youtube2 forced id option. I will look into that. I'm seeing yearly seasons being generated as I have things configured now, but they're easy enough to flatten in the library settings so they don't really affect me.

I just did some more searching and saw the SW_YOUTUBE_DATE switch in the code. If that enables generating the episode order based on the date stored in the filename, that may work to resolve my issue as well. I will try that next if something breaks. So far, the change I made appears to be working for me, but I am curious as to what will happen the next time the playlists are updated.

Thanks for the suggestion!

**Comment by ZeroQI** [comment]
Lines 864-925
SW_YOUTUBE_DATE is False by default so it sort using file date, if set to True it sorts using natural algorithm, so xxx 1, 2, … , 11 will be sorted as it should in titles

**Comment by ZeroQI** [comment]
Any feedback with youtube2 mode?

**Comment by crxm** [comment]
I wasn't able to find much information on youtube2 mode. I did want to try it, however, and it looked like all I needed to do to activate it was prepend 'youtube2-' before the channel/playlist ID in the parent folder, so that's what I did. One of my channel folders looks like this now:

[youtube2-UCFhXFikryT4aFcLkLw2LBLA]

Is that correct? I also enabled the SW_YOUTUBE_DATE switch, as sorting episodes by the upload date I have in my file names should really be all that I need for my purposes.

With both of those changes, the sorting of my Channel files looked great. They are still split into yearly seasons, but the episode numbers are cumulative. Season 2021 starts at E200-something, for example. I actually like this a lot as it is easier to navigate.

My Playlist files were still ordered by their playlist ranking rather than their upload date, however, so I'm still using my original change to reverse the ranking.
